### PR TITLE
Allow empty output nodes and measures at circuit level

### PR DESCRIPTION
**Pull request by thierry-martinez**
This pull request enables measures to be expressed at circuit level, in order to be able to consider circuits with classical outputs. Patterns are now allowed to measure all their nodes, and have an empty output node set, to enable to consider patterns where only classical outputs are considered.

Transpiled circuits can now have "measure" gates, introduced with the `circ.m(qubit, plane, angle)` method. The measured qubit cannot be used in any subsequent gate. The transpiler now returns a `TranspileResult` dataclass: the pattern is available in the `pattern` field, and the field `classical_outputs` contains the index where the classical measures can be found in the `results` array of the simulator.

### CODE REVIEW

File: tests/test_tnsim.py
Comment (shinich1) [inline tests/test_tnsim.py:1]: should we also test the `measure` feature for TN and DM backends?

**Review by shinich1 — COMMENTED** [review]
Could you also add appropriate lines to docs (e.g. `generator.rst`) so they show up in module references? also it might be good to illustrate the use of this feature in examples - perhaps simple example like `rotation.py`?

**Comment by shinich1** [comment]
example and docs bits can be done on my side as well, I will approve as soon as there's test for TN backend.

File: tests/test_tnsim.py
Comment (thierry-martinez) [reply tests/test_tnsim.py:1]: Done in aee982d. Thanks!

**Comment by thierry-martinez** [comment]
> Could you also add appropriate lines to docs (e.g. `generator.rst`) so they show up in module references? also it might be good to illustrate the use of this feature in examples - perhaps simple example like `rotation.py`?

I tried to improve the documentation in 1e97cdc. Thanks!

**Review by shinich1 — APPROVED** [review]
@thierry-martinez Great! please squash and merge.

**Comment by thierry-martinez** [comment]
Thanks! Merged.

**Pull request merged**